drools-ansible-rulebook-integration-api/src/main/java/org/drools/ansible/rulebook/integration/api/rulesengine/RulesSetEventStructure.java

```diff
@@ -0,0 +1,308 @@
+    }
+
+    /**
+     * normalize the event path to this format.
+     * person["name"] -> person.name
+     * person[0] -> person[]
+     * TODO: keeping the index person[0] might be better for the future validation.
+     */
+    private List<String> splitEventPath(String eventPath) {
+        String[] nodes = eventPath.split("\\.");
+        List<String> pathNodeList = new ArrayList<>();
+        for (String node : nodes) {
+            // handle bracket notation
+            int leftBracket = node.indexOf('[');
+            if (leftBracket > 0) {
+                int rightBracket = node.indexOf(']');
+                String body = node.substring(0, leftBracket);
+                String key = node.substring(leftBracket + 1, rightBracket);
+                if (isQuotedString(key)) {
+                    // String key means it's a map access. Equivalent to property
+                    key = key.substring(1, key.length() - 1);
+                    pathNodeList.add(body);
+                    pathNodeList.add(key);
+                } else if (isInteger(key)) {
+                    // Integer key means it's a list.
+                    pathNodeList.add(body + "[]");
+                } else {
+                    LOG.warn("Unknown key type: {} in the node {}", key, node);
+                    // return the path before the node
+                    return pathNodeList;
+                }
+
+                // TODO: handle nested bracket pairs. e.g. event.asd["x"][1][2]
+                if (rightBracket < node.length() - 1) {
+                    LOG.info("Multiple bracket pairs. {}", node);
+                    // return the path before the unparsed bracket
+                    return pathNodeList;
+                }
+            } else {
+                pathNodeList.add(node);
+            }
+        }
+        return pathNodeList;
+    }
+
+    private static boolean isQuotedString(String key) {
+        return key.length() >= 2 && (
+                (key.charAt(0) == '"' && key.charAt(key.length() - 1) == '"')
+                        ||
+                        (key.charAt(0) == '\'' && key.charAt(key.length() - 1) == '\'')
+        );
+    }
+
+    private static boolean isInteger(String str) {
+        if (str == null || str.isEmpty()) {
+            return false;
+        }
+        int i = 0;
+        if (str.charAt(0) == '-') {
+            if (str.length() == 1) {
+                return false;
+            }
+            i = 1;
+        }
+        for (; i < str.length(); i++) {
+            if (!Character.isDigit(str.charAt(i))) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    public boolean isValidated() {
+        return validated;
+    }
+
+    /**
+     * Validate the event path in RuleSets using the incoming event structure.
+     * Assume that the incoming event has the correct structure.
+     */
+    public void validate(String incomingJson) {
+        validated = true;
```

Comment:
If I understand correctly now this validation is performed only once per session which is probably different from what has been required. To implement this correctly I believe that we should have a Set of already emitted warning and avoid duplication through it, but of course this will introduce an even bigger performance cost :(

Reply:
@mariofusco Hmm, let me check with Madhu in slack.